### PR TITLE
undo transferFrom, assert balance in offRamp

### PR DESCRIPTION
**Pull request by RensR**
## Motivation
It is significantly cheaper to assert balances than do the transferFrom

## Solution
Undo recent approve changes, add balance assertions

had to undo some of the multiOfframp calldata changes to make it fit

### CODE REVIEW

File: contracts/src/v0.8/ccip/test/offRamp/EVM2EVMMultiOffRamp.t.sol [hunk 1]
```diff
@@ -2337,6 +2336,72 @@ contract EVM2EVMMultiOffRamp__releaseOrMintSingleToken is EVM2EVMMultiOffRampSet
...
+    vm.mockCallRevert(
+      s_destTokenBySourceToken[token], abi.encodeWithSelector(IERC20.balanceOf.selector, OWNER), revertData
+    );
+
+    vm.expectRevert(abi.encodeWithSelector(EVM2EVMMultiOffRamp.TokenHandlingError.selector, revertData));
+
+    s_offRamp.releaseOrMintSingleToken(tokenAmount, abi.encode(OWNER), OWNER, SOURCE_CHAIN_SELECTOR, "");
+  }
+
+  function test_releaseOrMintToken_ReleaseOrMintBalanceMismatch_Revert() public {
```
Comment (elatoskinas) [inline contracts/src/v0.8/ccip/test/offRamp/EVM2EVMMultiOffRamp.t.sol:2385]: suggestion: we should add a test where `expectedAmount < actualAmount` (it should trigger an overflow)

**Review by elatoskinas — APPROVED** [review]

File: contracts/src/v0.8/ccip/test/offRamp/EVM2EVMMultiOffRamp.t.sol (hunk 1 above)
Comment (RensR) [reply contracts/src/v0.8/ccip/test/offRamp/EVM2EVMMultiOffRamp.t.sol:2385]: That requires quite a bit more setup, but probably worth a test

File: contracts/src/v0.8/ccip/offRamp/EVM2EVMMultiOffRamp.sol [hunk 2]
```diff
@@ -855,21 +861,39 @@ contract EVM2EVMMultiOffRamp is ITypeAndVersion, MultiOCR3Base {
...
+      token,
+      gasLimit,
       Internal.GAS_FOR_CALL_EXACT_CHECK,
       Internal.MAX_RET_BYTES
     );
-
     if (!success) revert TokenHandlingError(returnData);
 
-    return Client.EVMTokenAmount({token: localToken, amount: localAmount});
+    // If the call was successful, the returnData should contain only the pre-balance.
```
Comment (RayXpub) [inline contracts/src/v0.8/ccip/offRamp/EVM2EVMMultiOffRamp.sol]: nit: we also use this for the post-balance, maybe update comment to:
```
// If the call was successful, the returnData should contain only the balance.
```

File: contracts/src/v0.8/ccip/pools/BurnMintTokenPoolAbstract.sol [hunk 3]
```diff
@@ -40,8 +40,7 @@ abstract contract BurnMintTokenPoolAbstract is TokenPool {
     _validateReleaseOrMint(releaseOrMintIn);
 
     // Mint to the offRamp, which forwards it to the recipient
```
Comment (RayXpub) [inline contracts/src/v0.8/ccip/pools/BurnMintTokenPoolAbstract.sol]:
```suggestion
    // Mint to the recipient
```

File: contracts/src/v0.8/ccip/pools/LockReleaseTokenPool.sol [hunk 4]
```diff
@@ -68,7 +68,7 @@ contract LockReleaseTokenPool is TokenPool, ILiquidityContainer, ITypeAndVersion
     _validateReleaseOrMint(releaseOrMintIn);
 
     // Release to the offRamp, which forwards it to the recipient
```
Comment (RayXpub) [inline contracts/src/v0.8/ccip/pools/LockReleaseTokenPool.sol]:
```suggestion
    // Release to the recipient
```

File: contracts/src/v0.8/ccip/pools/LockReleaseTokenPoolAndProxy.sol [hunk 5]
```diff
@@ -71,7 +71,7 @@ contract LockReleaseTokenPoolAndProxy is LegacyPoolWrapper, ILiquidityContainer,
 
     if (!_hasLegacyPool()) {
       // Release to the offRamp, which forwards it to the recipient
```
Comment (RayXpub) [inline contracts/src/v0.8/ccip/pools/LockReleaseTokenPoolAndProxy.sol]:
```suggestion
      // Release to the recipient
```

File: contracts/scripts/native_solc_compile_all_ccip [hunk 6]
```diff
@@ -28,10 +28,10 @@ compileContract () {
   local optimize_runs=$OPTIMIZE_RUNS
 
   case $1 in
-    "ccip/offRamp/EVM2EVMOffRamp.sol")
-    echo "OffRamp uses $OPTIMIZE_RUNS_OFFRAMP optimizer runs."
-      optimize_runs=$OPTIMIZE_RUNS_OFFRAMP
```
Comment (RensR) [inline contracts/scripts/native_solc_compile_all_ccip:33]: I believe we accidentally compiled the offramp with 2k opts instead of the 22k now.

**Review by RayXpub — APPROVED** [review]

**Review by RayXpub — APPROVED** [review]
